### ISSUE TITLE
support remote signing via reproducible builds

### ISSUE THREAD

**Issue by JanZerebecki** [sigstore/cosign#2049]
**Description**

Have cosign on the developers machine run the same build, creating a signature locally and then sending the signature to the remote in machine, where it is checked and applied. (See also https://www.qubes-os.org/doc/split-gpg/ for something similar in concept.)
Besides allowing signing on an untrusted machine while still retaining security, this makes it possible to independently verify the process without relying on the signature.

I have seen people using cosign with fulcio to sign builds in e.g. GitHub Actions. This would support signing in a GitHub action while affording better security.

**Comment by znewman01** [comment]
Overall, I agree with you that reproducible builds and software signatures go well together!

Is your argument something like the following?

- Some people are having GitHub Actions sign with private keys stored as GitHub Secrets.
- This is bad for a couple of reasons:
  - GitHub Actions can now sign anything it wants. So if (1) GitHub is compromised, or (2, more likely) the GitHub actions configuration changes, we could have signatures of evil builds.
  - That private key had to be generated somewhere else, and so it might have leaked as it was uploaded to GitHub.
- Instead, we should have developers build + sign locally, and then send that signature to GitHub Actions
  - GitHub Actions publishes the secret *only if* it got the same result.

Then, the overall effect is: *both* GitHub CI *and* the developer need to cooperate to publish an artifact. I think this is a good goal, and agree that reproducible builds help with this.

However, to me, it sounds more straightforward to have GitHub Actions sign the artifact as well, and then clients can check that there are signatures from *both* the developer and CI. Do you agree that the effect would be similar?

(If you're worried about accidentally publishing artifacts without appropriate signatures, you can have distinct CI jobs for (1) build+sign and (2) publish release, with the latter gated on the same threshold of signatures.)

But I think things get even better with [keyless signatures](https://github.com/sigstore/cosign/blob/dad7d3f1892a9c00c0bf5bd09eefe1a4ad13218d/KEYLESS.md). Then, we can do the following:

- I, the developer, build and sign my artifact using a key I only hold locally.
- From the public source, GitHub CI builds a (bit-for-bit identical) artifact, and signs using a Sigstore signature based on the [GitHub Actions OIDC identity](https://docs.github.com/en/actions/deployment/security-hardening-your-deployments/about-security-hardening-with-openid-connect).
- Now, as a client, I can check that I have a signature from *both* the developer *and* GitHub CI.
  - Bonus points for using [in-toto](https://in-toto.io/) to manage this threshold. It's capable of expressing things like: "I need a claim that the artifact with hash X was built from commit Y using build process Z, signed by (GitHub CI AND (developer A OR developer B))" 


---------------------------------------


Another question: I don't see why this is a cosign feature. The reproducibility check would look something like:

```
$ gcc -o a.out main.c
$ [get the signature from the dev]
$ cosign verify-blob --signature user-sig.sig --key user-key.pub
$ [publish a.out as usual]
```

The tricky part lies in "get the signature from the user". How would cosign implement this feature? It seems to me easier to have the developer provide the signature as a build input when they kick off the release process.

------------------------------------------------
  
Curious to hear your thoughts!

**Comment by JanZerebecki** [comment]
Yes, my argument was somewhat like that. Important details in my scenario are that they are 1) using Github Web UI to do merges (not git merge and push) and tags 2) do not use commit signing 3) already use what cosign calls keyless mode.

So (3) makes it different, but your criticism of your scenario applies equally to mine as the entity in control of the commit signing key used for Web UI actions, any secret in the CI, and the OIDC endpoint are the same.

### Simplifying while slightly improving security

Both a secret key in the CI or OIDC can help with associating the target project and branch of a merge or target project and tag name. However for software projects released from git with tags that are machine comparable versions, we do not need to differentiate the project/repo if the user is differentiated. Because the source branch and user are already included in the signed commit or tag created by the back end of the Web UI, we can write the public key identifier, hash of user, and tag hash to rekor. If rekor were to implement a solution to split view and availability, that would give us without any additional authentication even better security.

We can do better than (3) with reproducible builds and authenticating the input via the signature created by (1). This as a side effect made the process easier on the person making the release. The CI can ensure this by doing the build twice.

### Aside on keys

We could improve the security quite a bit by making git sign commits and tags by default. It would create a key on first invocation without any user input. This would remove all the friction, that GPG singing commits usually involves.

Github has no key rotation in place for the GPG key the Web UI uses to sign. (Same for their SSH host key.) And they only use one secret across all involved servers.

I have not checked the details of OIDC, does it mandate something like key rotation chained with signatures and allow offline verification after the fact?

To get short lived keys we can use something like is done in Decentralised IDentifieres to chain key rotations and do that automatically during signing if x time went by. For developers with multiple machines, we can offer cross signing. For recovery we can store a chained recovery use once key in their mailbox/dropbox/etc or use a passphrase to derive a key or use social recovery with a quorum. Maybe FIDO2 keys can be directly incorporated into such a scheme, to allow offline verifiable 2fa, or as the sole secret storage.

To not rely only on the secrecy of the private key we can use keys whose policy is to only make signatures that are published to rekor.

If we want to bind keys to Github identities, we can use something like what Keybase or https://github.com/wiktor-k/openpgp-proofs does.

We can define who owns a repo based on these keys in the repo itself, e.g. in a file named `.gittrust`. And have git verify by default if this is present in the default branch.

But this aside is not integral to this feature request.

### Strengthen the source

To avoid the centralized single point of failure, any maintainers from the project could then confirm the rekor entry based on their local git repo by something like `cd /path/repo && cosign confirm tag-name`. 3rd parties could review it with something like https://github.com/crev-dev/crev and store a reference to the review in rekor.

### Strengthen the binary

Anyone can run a rebuilder like https://github.com/in-toto/rebuilderd that follows the repo and/or the entries for the repo in rekor, reproduces the build and sends the signed attestation to rekor. Github CI could be one of those.

### Tying it together

I'm a bit unsure about this one as I'm not familiar with some of the parts affected.

We can combine the above two and a) satisfy legacy systems with just one signature that can not do any multi signature policy and not query rekor (otherwise we could just do with the above and a policy on how to verify from rekor and possibly gate the CI publish on a threshold appearing in rekor) and b) make sure the CI doesn't release without maintainer confirmation by: 1) In the Github Action send intent to sign with the intoto meta data which is enough to reproduce the build to rekor (this is so the developer doesn't accidentally sign something which was not tested to at least be reproducible by two runs) 2) wait for signature from next step to appear in rekor and then transplant it to the binary 3a) (this overlaps with what https://github.com/in-toto/rebuilderd does) on the developers machine trigger a build, receive the intoto meta data, check that it matches the input the developer provided (e.g. same git commit or tag provided locally), check that the build result is the same, 3b) sign it, send signature to rekor.

Using rekor here as the transport is just an example, it could as well be a git branch, that the CI can only append to.

### Make secure easy

I have the hunch that for good UI cosign should itself call whatever (in-toto-verify?) is needed to reproduce based on the in-toto recipe derived from local git which the CI also uses, as manual steps lead to things being non-reproducible.

**Comment by znewman01** [comment]
I'd like to check that I'm understanding your proposal: GitHub merge commits are signed by GitHub, *not* maintainers, so if we check (1) that a commit is signed, and (2) a build is done by GitHub CI, (1) isn't giving us a whole lot of assurance (single-point-of-failure, no tie to maintainer identity).

But we like the GitHub workflow; how do we get the security assurance of signed commits without giving that up? We can have maintainers add extra signatures after-the-fact, after reproducing the build from their local checkout.

I think there are a bunch of good ideas here, but it's a little difficult to understand what concretely needs to get done.  Perhaps the best thing to do would be to build a *separate* tool? Or maybe this is a feature-request for the in-toto tooling itself?

Some alternatives here include [Guix's approach](https://programming-journal.org/2023/7/1/), which doesn't quite work in a GitHub-centric workflow but is interesting enough that I'll include it.

**Comment by znewman01** [comment]
Going through some old issues here. I still think that this issue contains good ideas, but I'm skeptical that Cosign is the right place to implement them. in-toto might make more sense. CC @adityasaky @SantiagoTorres

**Comment by SantiagoTorres** [comment]
I think there are things to be said around using in-toto + cosign to sign for a rebuild result. We are already able to attest to anything w/ cosign, so I wonder what else is there to do. Let me take a better look at the ctx and think of a path fw.

**Comment by adityasaky** [comment]
It makes sense to me to record results of every build using in-toto (we already do this in rebuilderd) and each builder (whether GitHub actions or the developer) can use OIDC identities to sign the metadata (https://github.com/in-toto/ITE/blob/master/ITE/7/README.adoc makes this possible in in-toto). So for the proposed workflow, we'd either use in-toto + Fulcio ITE-7 bits to sign the metadata or have cosign generate an in-toto attestation (already possible) for the built artifact signed with a Fulcio key. I think the only necessary change to an in-toto implementation is the Fulcio integration. Is that right?

**Comment by JanZerebecki** [comment]
Thank you for your responses.

To improve the source part I started a proof of concept https://gitlab.com/source-security/git-verify .

OIDC doesn't seem to a be good fit for verifying code signatures as typical security requirements are out of scope and https://docs.sigstore.dev/security#fulcio-security-model even deems further ones out of scope. But we should probably discuss this in the Fulcio project, not here.

Maybe for cosign I'm only missing changes to its documentation, reading the README.md of cosign doesn't mention anything about reproducing a build. It mentions Fulcio very prominently instead. Though it mentions in-toto, but not how you get to that point. The detailed documentation says how to upload a signature you got from elsewhere, but not why that is useful or when you want to chose that over alternatives.

**Comment by znewman01** [comment]
> To improve the source part I started a proof of concept https://gitlab.com/source-security/git-verify .

Very cool! I'll be sure to watch that as it progresses. If you're interested in using Sigstore as a building block and it's missing features, feel free to open issues and we'll try to help out.

> Maybe for cosign I'm only missing changes to its documentation, reading the README.md of cosign doesn't mention anything about reproducing a build. Though it mentions in-toto, but not how you get to that point. The detailed documentation says how to upload a signature you got from elsewhere, but not why that is useful or when you want to chose that over alternatives.

In general, I don't think we can capture every use case in the README. IMO it's sufficient to provide Cosign as a primitive and describe how it works.

That said, if you're interested in writing something up about this, we could publish it on the [Sigstore blog](https://blog.sigstore.dev/), and we're happy to promote projects that *use* Cosign and Sigstore in general.
